candy-machine/program/src/processor/freeze/thaw_nft.rs

```diff
@@ -0,0 +1,107 @@
+    token_account: Account<'info, TokenAccount>,
+    /// CHECK: checked in token_account constraints
+    owner: UncheckedAccount<'info>,
+    mint: Account<'info, Mint>,
+    /// CHECK: account checked in CPI
+    edition: UncheckedAccount<'info>,
+    #[account(mut)]
+    payer: Signer<'info>,
+    token_program: Program<'info, Token>,
+    /// CHECK: checked in account constraints
+    #[account(address = mpl_token_metadata::id())]
+    token_metadata_program: UncheckedAccount<'info>,
+    system_program: Program<'info, System>,
+}
+
+pub fn handle_thaw_nft(ctx: Context<ThawNFT>) -> Result<()> {
+    let freeze_pda = &mut ctx.accounts.freeze_pda;
+    let candy_machine = &mut ctx.accounts.candy_machine;
+    let current_timestamp = Clock::get()?.unix_timestamp;
+    let can_thaw = if candy_machine.data_is_empty() {
+        true
+    } else {
+        let candy_struct: Account<CandyMachine> =
+            Account::try_from(&candy_machine.to_account_info())?;
+        freeze_pda.thaw_eligible(current_timestamp, &candy_struct)
+    };
+    msg!("Can thaw: {}", can_thaw);
+    if !can_thaw {
+        return err!(CandyError::InvalidThawNft);
+    }
+    let token_account = &ctx.accounts.token_account;
+    let mint = &ctx.accounts.mint;
+    let edition = &ctx.accounts.edition;
+    let payer = &ctx.accounts.payer;
+    let owner = &ctx.accounts.owner;
+    let token_program = &ctx.accounts.token_program;
+    let token_metadata_program = &ctx.accounts.token_metadata_program;
+    let freeze_seeds = [
+        FreezePDA::PREFIX.as_bytes(),
+        candy_machine.key.as_ref(),
+        &[*ctx.bumps.get("freeze_pda").unwrap()],
+    ];
+    if token_account.is_frozen() {
+        msg!("Token account is frozen! Now attempting to thaw!");
+        invoke_signed(
+            &thaw_delegated_account(
+                mpl_token_metadata::ID,
+                freeze_pda.key(),
+                token_account.key(),
+                edition.key(),
+                mint.key(),
+            ),
+            &[
+                freeze_pda.to_account_info(),
+                token_account.to_account_info(),
+                edition.to_account_info(),
+                mint.to_account_info(),
+                token_program.to_account_info(),
+                token_metadata_program.to_account_info(),
+            ],
+            &[&freeze_seeds],
+        )?;
+        freeze_pda.frozen_count = freeze_pda.frozen_count.saturating_sub(1);
+        if freeze_pda.freeze_fee > 0 && freeze_pda.frozen_count > 0 {
+            transfer(
+                CpiContext::new(
+                    ctx.accounts.system_program.to_account_info(),
+                    Transfer {
+                        from: freeze_pda.to_account_info(),
+                        to: payer.to_account_info(),
+                    },
+                ),
+                freeze_pda.freeze_fee,
+            )?;
+        }
+    } else {
+        msg!("Token account is not frozen!");
```

Comment:
We haven't been doing bot taxes on non mint commands. I think what I was doing was allowing this handler to always be run from both the owner and a random other party. The issue is if a random party runs it they cant revoke delegate authority (only owner can do this). So you can run it again and it would basically just revoke.